### PR TITLE
Dịch "encoder-decoder_vn" - Phần 1

### LINKED ISSUE #1690: Dịch "encoder-decoder_vn" - Phần 1
Phần cần dịch: https://github.com/aivivn/d2l-vn/blame/master/chapter_recurrent-modern/encoder-decoder_vn.md#L1-L90

### Hướng dẫn
- [Dịch trực tiếp trên website](https://github.com/aivivn/d2l-vn/blob/master/CONTRIBUTING.md).
- Sau khi nhấn vào phần cần dịch ở trên, hãy chọn **Normal view** và làm theo hướng dẫn ở trên để có thể bắt đầu dịch.

### PR DESCRIPTION
**Pull request by nguyenhoa93**
## Pull Request Checklist

**Đánh dấu tick (sau khi "Create Pull Request") vào các mục dưới đây:**

* [x] Pull Request này tương ứng với issue nào? Trả lời số issue sau dấu `#` (không có dấu cách): Close #1690

* [x] Bản dịch này có bám sát **[Bảng thuật ngữ](https://github.com/aivivn/d2l-vn/blob/master/glossary.md)** không? Nếu một từ/cụm chưa có trong Bảng thuật ngữ, hãy tạo một Pull Request riêng đề xuất cách dịch từ/cụm đó.

* [x] Format của bản gốc có được giữ nguyên không?
    * Các phần in nghiêng, in đậm.
    * Tiêu đề (số lượng dấu `#` đầu dòng).
    * Bảng biểu, chú thích cho bảng (dòng phía trên mỗi bảng bắt đầu bằng dấu `:`).
    * Dịch các từ trong hình vẽ nếu cần. Các dòng có hình có dạng: `caption`.
    * Dịch các chú thích hình vẽ (thay các cụm `*dịch chú thích ảnh phía trên*` bằng bản dịch tiếng Việt).
    * Không dịch các phần code (nằm giữa hai dấu `````).
    * Copy các công thức toán từ bản gốc (các đoạn có `$`).
    * Giữ các dòng gán nhãn (bắt đầu với `:label:`).
    * Không tự thêm bớt các dòng trắng.
     
* [x] Trong một đoạn văn, mỗi câu văn đã được viết trong một dòng, giữa các dòng không có dòng trắng. Markdown sẽ coi những dòng liền nhau không có dòng trắng là một đoạn văn. Việc này giúp công đoạn review được thuận tiện hơn.

* [ ] Tất cả chữ trong các hình vẽ đã được dịch. Nếu chưa, vui lòng xem thêm **[hướng dẫn hoặc yêu cầu dịch hình](https://github.com/aivivn/d2l-vn/blob/master/img/README.md)**.

* [x] Điền tên của bạn vào mục "Những người thực hiện" ở cuối file.

* [ ] **Sau khi được approve**, điền tên những reviewers mà bạn thấy có nhiều đóng góp cho Pull Request này. Xem danh sách Contributors **[tại đây](https://github.com/aivivn/d2l-vn/blob/master/docs/contributors_info.md)**.

## Với Reviewer

1. Chỉ review nếu Pull Request này đã sẵn sàng cho việc review.

2. Đã đọc kỹ hướng dẫn đóng góp và xem qua **[Bảng thuật ngữ](https://github.com/aivivn/d2l-vn/blob/master/glossary.md)**.

3. **Bạn có thể đề xuất thay đổi cách dịch mỗi dòng trực tiếp như trong hình dưới đây:**
img

### CODE REVIEW

File: chapter_recurrent-modern/encoder-decoder_vn.md [hunk 1]
```diff
@@ -16,33 +16,37 @@ In machine translation, the encoder transforms a source sentence, e.g., "Hello w
 The decoder then uses this state to generate the translated target sentence, e.g., "Bonjour le monde.".
 -->
 
-*dịch đoạn phía trên*
+*Kiến trúc mã hoá - giải mã* là một thiết kế mạng nơ-ron.
+Như đã được trình bày trong :numref:`fig_encoder_decoder`, kiến trúc này được phân làm 2 phần: bộ mã hoá và bộ giải mã.
+Bộ mã hoá đóng vai trò mã hoá đầu vào thành trạng thái, trạng thái thường chứa vài tensor.
+Tiếp đó, trạng thái được truyền vào bộ giải mã để sinh đầu ra.
+Trong dịch máy, bộ mã hoá biến đổi một câu nguồn, ví dụ như "Hello world.", thành trạng thái, chẳng hạn là một vector, mà có thể nắm bắt được thông tin ngữ nghĩa của câu đó.
```
Comment (cuongvng) [inline chapter_recurrent-modern/encoder-decoder_vn.md:24]:
```suggestion
Sau đó bộ giải mã sử dụng trạng thái này để dịch câu sang ngôn ngữ đích, ví dụ sang tiếng Pháp "Bonjour le monde.".
```
Thiếu 1 câu nhé.

File: chapter_recurrent-modern/encoder-decoder_vn.md [hunk 2]
```diff
@@ -16,33 +16,37 @@ In machine translation, the encoder transforms a source sentence, e.g., "Hello w
...
-![*dịch chú thích ảnh phía trên*](../img/encoder-decoder.svg)
+![Kiến trúc mã hoá - giải mã](../img/encoder-decoder.svg)
 :label:`fig_encoder_decoder`
 
 <!--
 In this section, we will show an interface to implement this encoder-decoder architecture.
 -->
 
-*dịch đoạn phía trên*
+Trong phần này, chúng tô sẽ trình bày cách để thực hiện kiến trúc mã hoá - giải mã.
```
Comment (cuongvng) [inline chapter_recurrent-modern/encoder-decoder_vn.md:36]:
```suggestion
Phần này trình bày khuôn mẫu (*interface*) để lập trình kiến trúc mã hoá - giải mã.
```

File: chapter_recurrent-modern/encoder-decoder_vn.md [hunk 3]
```diff
@@ -70,7 +74,9 @@ In the forward method, the decoder takes both inputs, e.g., a target sentence an
 It returns outputs, with potentially modified state if the encoder contains RNN layers.
 -->
 
-*dịch đoạn phía trên*
+Bộ giải mã có một phương thức bổ sung gọi là `init_state` nhằm phân tích đầu ra của bộ mã hoá với những thông tin có thể được bổ sung, ví dụ như độ dài cho phép của đầu vào, để đưa ra trạng thái mà nó cần.
+Trong lan truyền thuận, bộ giải mã nhận hai đầu vào, ví dụ là một câu đích và trạng thái.
```
Comment (cuongvng) [inline chapter_recurrent-modern/encoder-decoder_vn.md:78]:
```suggestion
Trong lan truyền xuôi, bộ giải mã nhận hai đầu vào, ví dụ như một câu đích và trạng thái.
```

Comment (nguyenhoa93) [reply chapter_recurrent-modern/encoder-decoder_vn.md:78]: Cảm ơn, mình đã check lại glossary thì đúng là đã thống nhất là `lan truyền xuôi` thay vì là `lan truyền thuận`.

File: chapter_recurrent-modern/encoder-decoder_vn.md [hunk 4]
```diff
@@ -16,33 +16,37 @@ In machine translation, the encoder transforms a source sentence, e.g., "Hello w
 The decoder then uses this state to generate the translated target sentence, e.g., "Bonjour le monde.".
 -->
 
-*dịch đoạn phía trên*
-
+*Kiến trúc mã hoá - giải mã* là một khuôn mẫu thiết kế mạng nơ-ron.
+Như trình bày trong :numref:`fig_encoder_decoder`, kiến trúc này có 2 phần: bộ mã hoá và bộ giải mã.
+Bộ mã hoá đóng vai trò mã hoá đầu vào thành trạng thái, trạng thái thường chứa vài tensor.
```
Comment (lkhphuc) [inline chapter_recurrent-modern/encoder-decoder_vn.md:21]:
```suggestion
Bộ mã hoá đóng vai trò mã hoá đầu vào thành trạng thái, thường bao gồm một vài tensor.
```

File: chapter_recurrent-modern/encoder-decoder_vn.md [hunk 5]
```diff
@@ -16,33 +16,37 @@ In machine translation, the encoder transforms a source sentence, e.g., "Hello w
...
-![*dịch chú thích ảnh phía trên*](../img/encoder-decoder.svg)
+![Kiến trúc mã hoá - giải mã](../img/encoder-decoder.svg)
 :label:`fig_encoder_decoder`
 
 <!--
 In this section, we will show an interface to implement this encoder-decoder architecture.
 -->
 
-*dịch đoạn phía trên*
+Phần này trình bày khuôn mẫu (*interface*) để lập trình kiến trúc mã hoá - giải mã.
```
Comment (lkhphuc) [inline chapter_recurrent-modern/encoder-decoder_vn.md:36]:
```suggestion
Phần này sẽ trình bày một giao diện để lập trình kiến trúc mã hoá - giải mã.
```

File: chapter_recurrent-modern/encoder-decoder_vn.md [hunk 6]
```diff
@@ -70,7 +74,9 @@ In the forward method, the decoder takes both inputs, e.g., a target sentence an
 It returns outputs, with potentially modified state if the encoder contains RNN layers.
 -->
 
-*dịch đoạn phía trên*
+Bộ giải mã có một phương thức bổ sung gọi là `init_state` nhằm phân tích đầu ra của bộ mã hoá với những thông tin có thể được bổ sung, ví dụ như độ dài cho phép của đầu vào, để đưa ra trạng thái mà nó cần.
```
Comment (lkhphuc) [inline chapter_recurrent-modern/encoder-decoder_vn.md:77]:
```suggestion
Bộ giải mã có một phương thức bổ sung gọi là `init_state` nhằm phân tích đầu ra của bộ mã hoá với những thông tin có thể được bổ sung, ví dụ như độ dài hợp lệ của đầu vào, để đưa ra trạng thái mà nó cần.
```

File: chapter_recurrent-modern/encoder-decoder_vn.md [hunk 7]
```diff
@@ -70,7 +74,9 @@ In the forward method, the decoder takes both inputs, e.g., a target sentence an
...
+Trong lan truyền xuôi, bộ giải mã nhận hai đầu vào, ví dụ như một câu đích và trạng thái.
```
Comment (lkhphuc) [inline chapter_recurrent-modern/encoder-decoder_vn.md:78]:
```suggestion
Trong hàm truyền xuôi, bộ giải mã nhận hai đầu vào, ví dụ như một câu cần dịch và trạng thái.
```
chỗ này bị nhầm thì phải, ở trên ghi `target sentence` tiếng Pháp là đầu ra của decoder mà.

Comment (nguyenhoa93) [reply chapter_recurrent-modern/encoder-decoder_vn.md:78]: @lkhphuc 
Mình không nghĩ vậy đâu. Trong quá trình **training** decoder có thể nhận đầu vào là embedding của câu đích (ví dụ: trong seq2seq).
seq2seq_details

(https://d2l.ai/chapter_recurrent-modern/seq2seq.html)
```
def train_s2s_ch9(model, data_iter, lr, num_epochs, ctx):
    model.initialize(init.Xavier(), force_reinit=True, ctx=ctx)
    trainer = gluon.Trainer(model.collect_params(),
                            'adam', {'learning_rate': lr})
    loss = MaskedSoftmaxCELoss()
    animator = d2l.Animator(xlabel='epoch', ylabel='loss',
                            xlim=[1, num_epochs], ylim=[0, 0.25])
    for epoch in range(1, num_epochs + 1):
        timer = d2l.Timer()
        metric = d2l.Accumulator(2)  # loss_sum, num_tokens
        for batch in data_iter:
            X, X_vlen, Y, Y_vlen = [x.as_in_ctx(ctx) for x in batch]
            Y_input, Y_label, Y_vlen = Y[:, :-1], Y[:, 1:], Y_vlen-1
            with autograd.record():
                Y_hat, _ = model(X, Y_input, X_vlen, Y_vlen)
                l = loss(Y_hat, Y_label, Y_vlen)
            l.backward()
            d2l.grad_clipping(model, 1)
            num_tokens = Y_vlen.sum()
            trainer.step(num_tokens)
            metric.add(l.sum(), num_tokens)
        if epoch % 10 == 0:
            animator.add(epoch, (metric[0]/metric[1],))
    print('loss %.3f, %d tokens/sec on %s ' % (
        metric[0]/metric[1], metric[1]/timer.stop(), ctx))
```
`model` ở đây là: (https://d2l.ai/chapter_recurrent-modern/encoder-decoder.html)
```
class EncoderDecoder(nn.Block):
    """The base class for the encoder-decoder architecture."""
    def __init__(self, encoder, decoder, **kwargs):
        super(EncoderDecoder, self).__init__(**kwargs)
        self.encoder = encoder
        self.decoder = decoder

    def forward(self, enc_X, dec_X, *args):
        enc_outputs = self.encoder(enc_X, *args)
        dec_state = self.decoder.init_state(enc_outputs, *args)
        return self.decoder(dec_X, dec_state)
```

`Y_hat, _ = model(X, Y_input, X_vlen, Y_vlen)` thì argument `Y_input` ở đây tương ứng với `dec_X` (đầu vào của decoder). Y_input chính là embeding của câu đích vì:
```
for batch in data_iter:
            X, X_vlen, Y, Y_vlen = [x.as_in_ctx(ctx) for x in batch]
            Y_input, Y_label, Y_vlen = Y[:, :-1], Y[:, 1:], Y_vlen-1
```
mà `data_inter` được generate ra bằng hàm này: https://github.com/d2l-ai/d2l-en/blob/master/d2l/mxnet.py#L786-L799

File: chapter_recurrent-modern/encoder-decoder_vn.md (hunk 4 above)
Comment (nguyenhoa93) [reply chapter_recurrent-modern/encoder-decoder_vn.md:21]: Mình nghĩ khi viết `Bộ mã hoá đóng vai trò mã hoá đầu vào thành trạng thái, thường bao gồm một vài tensor.` có thể người đọc sẽ hiểu rằng bộ mã hoá bao gồm một vài tensor.

File: chapter_recurrent-modern/encoder-decoder_vn.md (hunk 7 above)
Comment (lkhphuc) [reply chapter_recurrent-modern/encoder-decoder_vn.md:78]: Thank @nguyenhoa93, đúng là mình nhầm. Decoder nhận đầu vào là encoder state embedding và câu đích từ đầu tới trước vị trí cần dịch. 
Như vậy `Y_input` là text thôi chứ không phải embedding đâu nhé.
